plugins/trace_api_plugin/include/eosio/trace_api_plugin/store_provider.hpp

```diff
@@ -0,0 +1,186 @@
+#pragma once
+
+#include <ios>
+#include <fc/io/cfile.hpp>
+#include <boost/filesystem.hpp>
+#include <fc/variant.hpp>
+#include <eosio/trace_api_plugin/metadata_log.hpp>
+#include <eosio/trace_api_plugin/data_log.hpp>
+
+namespace eosio::trace_api_plugin {
+   using namespace boost::filesystem;
+
+   class path_does_not_exist : public std::runtime_error {
+   public:
+      explicit path_does_not_exist(const char* what_arg)
+         :std::runtime_error(what_arg)
+      {}
+      explicit path_does_not_exist(const std::string& what_arg)
+         :std::runtime_error(what_arg)
+      {}
+   };
+
+   class old_slice_version : public std::runtime_error {
+   public:
+      explicit old_slice_version(const char* what_arg)
+         :std::runtime_error(what_arg)
+      {}
+      explicit old_slice_version(const std::string& what_arg)
+         :std::runtime_error(what_arg)
+      {}
+   };
+
+   class incompatible_slice_files : public std::runtime_error {
+   public:
+      explicit incompatible_slice_files(const char* what_arg)
+         :std::runtime_error(what_arg)
+      {}
+      explicit incompatible_slice_files(const std::string& what_arg)
+         :std::runtime_error(what_arg)
+      {}
+   };
+
+   class malformed_slice_file : public std::runtime_error {
+   public:
+      explicit malformed_slice_file(const char* what_arg)
+         :std::runtime_error(what_arg)
+      {}
+      explicit malformed_slice_file(const std::string& what_arg)
+         :std::runtime_error(what_arg)
+      {}
+   };
+
+   class store_provider;
+
+
+   class slice_provider {
+   public:
+      struct index_header {
+         uint32_t version;
+      };
+      slice_provider(const boost::filesystem::path& slice_dir, uint32_t width);
+
+      uint32_t slice_number(uint32_t block_height) const {
+         return block_height / _width;
+      }
+
+      bool find_index_slice(uint32_t slice_number, bool append, fc::cfile& slice_file) const;
+      bool find_trace_slice(uint32_t slice_number, bool append, fc::cfile& slice_file) const;
+
+   private:
+      // returns true if slice is created
+      bool find_slice(const char* slice_prefix, uint32_t slice_number, fc::cfile& slice_file) const;
+
+      const boost::filesystem::path _slice_dir;
+      const uint32_t _width;
+
+      static constexpr uint32_t _current_version = 1;
+
+      static constexpr uint _slice_size = 10;
+      static constexpr uint _null_term_size = 1;
+      static constexpr const char* _trace_prefix = "trace_";
+      static constexpr const char* _trace_index_prefix = "trace_index_";
+      static constexpr uint _max_file_size = 12 + 10 + 1 + 10 + 4 + 1; // "trace_index_" + 10-digits + '-' + 10-digits + ".log" + null-char
```

Comment:
Rename to `_max_filename_size`

Reply:
We are using C++17 so you can do: 
`static constexpr uint _max_filename_size = std::char_traits<char>::length(_trace_index_prefix) + 10 + 1 + 10 + std::char_traits<char>::length(_ext) + 1;`

Reply:
Add a constexpr for `_trace_ext = ".log"`

Reply:
`_slice_size` and `_null_term_size` are not used.

Reply:
Many of these things are only used in `store_provider::find_slice` or as a direct input (based on whether this is an index or data log).  Can we move all of them into the implementation of a `private` method like `make_slice_filename` with a `bool` or `enum` indicating the type of log?

I think that increase in locality will help the next engineer when they are reading the code.

Reply:
I moved the variables to an anonymous namespace, does that do it?

Reply:
that helps the visibility.  putting them in the function (while some find it abhorrent) means that you can see the definition and only uses of those variables in the same window/terminal.  I'm not going to press it just clarifying the ask.